### PR TITLE
Project: S3779 LZW ColGroup | ASML

### PR DESCRIPTION
**Pull request by MasterBrain2000**
Group Project for the AMLS Module

### CODE REVIEW

**Comment by janniklinde** [comment]
Your code does not build (see error logs). Could you please fix those issues to have a prototype that compiles?

**Comment by janniklinde** [comment]
@MasterBrain2000 @m-ollka @Mancer1 could you please address the issues causing the tests to fail? Thanks

**Comment by Mancer1** [comment]
@janniklinde, all the build problems have been addressed.

**Comment by Mancer1** [comment]
@janniklinde, for code coverage, aren't tests considered part of it? We have some helper functions from the old DP algorithm, and we used to test them against the new algorithm. These DP helper functions are not used anywhere other than the tests we wrote.

**Comment by Mancer1** [comment]
@janniklinde, How much code coverage is accepted? We have some null functions that are required to be implemented but never called.

**Comment by janniklinde** [comment]
Generally, all functions that are expected to provide a working implementation should be tested, @Mancer1. Coverage also includes nested function calls, so you should be able to achieve high coverage by testing the publicly available functions of the compressed column group. Please use the tests for existing `AColGroup` implementations as a reference.

Also, please avoid reimplementing helper functionality that is already available in the test utilities, such as random matrix generation.

**Comment by Mancer1** [comment]
@janniklinde, we have improved the coverage to 95%. I hope that suffices

**Comment by Mancer1** [comment]
@janniklinde, we have already reformatted our files that we edited. The files that java Codestyle check is showing to reformat are those not part of our PR. 

```bash
rice@Arnor:~/MasterBrain2000-Project-S3779-LZW-ColGroup$  dev/format-changed.sh
All PR-edited Java lines are correctly formatted.
rice@Arnor:~/MasterBrain2000-Project-S3779-LZW-ColGroup$ git status
On branch main
Your branch is up to date with 'origin/main'.
```

So, should we leave it be, since we were working on a stale version of main?

**Comment by janniklinde** [comment]
You can simply run the following @Mancer1:
```
git fetch upstream
dev/format-changed.sh upstream/main
```

**Comment by Mancer1** [comment]
> You can simply run the following @Mancer1:
> 
> ```
> git fetch upstream
> dev/format-changed.sh upstream/main
> ```

@janniklinde, I believe I tried it before. It didn't affect anything.


````bash
rice@Arnor:~/MasterBrain2000-Project-S3779-LZW-ColGroup$ git fetch upstream
dev/format-changed.sh upstream/main
remote: Enumerating objects: 299, done.
remote: Counting objects: 100% (172/172), done.
remote: Compressing objects: 100% (90/90), done.
remote: Total 299 (delta 105), reused 82 (delta 82), pack-reused 127 (from 2)
Receiving objects: 100% (299/299), 262.07 KiB | 507.00 KiB/s, done.
Resolving deltas: 100% (122/122), completed with 16 local objects.
From https://github.com/apache/systemds
 * [new branch]            dependabot/github_actions/docker/login-action-4.6.0 -> upstream/dependabot/github_actions/docker/login-action-4.6.0
   7a0e5ca0ac..9dccbc3e60  main       -> upstream/main
All PR-edited Java lines are correctly formatted.
rice@Arnor:~/MasterBrain2000-Project-S3779-LZW-ColGroup$ git status
On branch main
Your branch is up to date with 'origin/main'.
```

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class PiecewiseLinearUtils {
+	/**
+	 * Utility methods for piecewise linear compression of matric columns supports compression used the segmented least
+	 * squares algorithm which is implemented with dynamic programming and a successive method, which puts all values in
+	 * a segment till the target loss is exceeded
+	 */
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java]: Comment not up to date, either remove or update and put above class definition

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+		// Regress per Segment
+		for(int seg = 0; seg < numSeg; seg++) {
+			final int segstart = breakpoints[seg];
+			final int segEnd = breakpoints[seg + 1];
+			final double[] line = regressSegment(column, segstart, segEnd);
+			slopes[seg] = line[0];
+			intercepts[seg] = line[1];
+		}
+		return new SegmentedRegression(breakpoints, slopes, intercepts);
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java:100]: These comments are unnecessary. I'd prefer a clear documentation comment for the function itself.

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+		final double slope = ab[0];
+		final double intercept = ab[1];
+
+		double sse = 0.0;
+		for(int i = start; i < end; i++) {
+			double err = column[i] - (slope * i + intercept);
+			sse += err * err;
+		}
+		return sse;
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java]: Can be removed (including the test)

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupFactory.java [hunk 4]
```diff
@@ -1078,6 +1083,46 @@ private static AColGroup compressLinearFunctional(IColIndex colIndexes, MatrixBl
...
+	 * This method is the entry point to compress a matrix with piecewise linear compression The first method uses a
+	 * segmented least squares with dynamic programming to compress the columns The second method uses a successive
+	 * compression method, which compares each values in linear time and checks if the targetloss exceeded
+	 *
+	 * @param colIndexes the column indices to compress
+	 * @param in         the input Matrixblock containing the data
+	 * @param cs         compression settings to define the target loss, which should be considered
+	 * @return a piecewise linear compressed column group
+	 */
+
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupFactory.java]: Remove line

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupFactory.java [hunk 5]
```diff
@@ -1078,6 +1083,46 @@ private static AColGroup compressLinearFunctional(IColIndex colIndexes, MatrixBl
...
+				.compressSuccessivePiecewiseLinear(column, cs);
+			breakpointsPerCol[col] = fit.getBreakpoints();
+			interceptsPerCol[col] = fit.getIntercepts();
+			slopesPerCol[col] = fit.getSlopes();
+
+		}
+		return ColGroupPiecewiseLinearCompressed.create(colIndexes, breakpointsPerCol, slopesPerCol, interceptsPerCol,
+			numRows);
+
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupFactory.java]: Incorrect if `cs.transposed = true`, ther e you need to iterate over rows

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 6]
```diff
@@ -0,0 +1,1443 @@
...
+			for(int j = 0; j < twoDimArray_lenght; j++) {
+				twoDimArray[i][j] = in.readDouble();
+			}
+		}
+		return twoDimArray;
+	}
+
+	public static ColGroupPiecewiseLinearCompressed read(DataInput in) throws IOException {
+		// read ColGroupType written by AColGroup.write()
+		in.readByte();
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: This byte is already read in ColGroupIO, so this `readByte()` should be removed

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 7]
```diff
@@ -0,0 +1,1443 @@
...
+	 *
+	 * @param colIndices        the column indices this group represents
+	 * @param breakpointsPerCol breakpoint indices per column
+	 * @param slopesPerCol      slope of each segment per column
+	 * @param interceptsPerCol  intercept of each segment per column
+	 * @param numRows           number of rows in the original matrix
+	 * @return a new ColGroupPiecewiseLinearCompressed instance
+	 * @throws IllegalArgumentException if breakpoints are invalid or arrays are inconsistent
+	 */
+
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Remove line

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 8]
```diff
@@ -0,0 +1,1443 @@
...
+		int[][] bpCopy = new int[numCols][];
+		double[][] slopeCopy = new double[numCols][];
+		double[][] interceptCopy = new double[numCols][];
+		// defensive copy to prevent external modification
+		for(int c = 0; c < numCols; c++) {
+			bpCopy[c] = Arrays.copyOf(breakpointsPerCol[c], breakpointsPerCol[c].length);
+			slopeCopy[c] = Arrays.copyOf(slopesPerCol[c], slopesPerCol[c].length);
+			interceptCopy[c] = Arrays.copyOf(interceptsPerCol[c], interceptsPerCol[c].length);
+		}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Is this defensive copy necessary?

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 9]
```diff
@@ -0,0 +1,1443 @@
...
+
+					if(dbRow >= 0 && dbRow < db.numRows() && dbCol >= 0 && dbCol < db.numCols()) {
+						db.set(dbRow, dbCol, yhat);
+					}
+				}
+
+			}
+
+		}
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:153]: Very defensive checks (e.g. `if(db == null ...)`, `if(dbRow >= 0)`...). Please remove unnecessary ones

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 10]
```diff
@@ -0,0 +1,1443 @@
...
+	 * @param r      row
+	 * @param colIdx column index in the _colIndexes.
+	 * @return reconstructed value with slope[segment]*r+intercepts[segment]
+	 */
+	@Override
+	public double getIdx(int r, int colIdx) {
+		// safety check
+		if(r < 0 || r >= numRows || colIdx < 0 || colIdx >= _colIndexes.size()) {
+			return 0.0;
+		}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Rather throw an error

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 11]
```diff
@@ -0,0 +1,1443 @@
...
+
+		for(int c = 0; c < numCols; c++) {
+			ret += (long) MemoryEstimates.intArrayCost(breakpointsPerCol[c].length);
+			ret += (long) MemoryEstimates.doubleArrayCost(slopesPerCol[c].length);
+			ret += (long) MemoryEstimates.doubleArrayCost(interceptsPerCol[c].length);
+		}
+
+		return ret;
+
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:230]: Looks incorrect, exact size on disk corresponds to serialized bytes (e.g. what is written into an output stream). For example reference pointers are not written to disk

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 12]
```diff
@@ -0,0 +1,1443 @@
...
+	/**
+	 * Computes the column sums of the decompressed matrix using sum of arithmetic series Where sumX = len * (2*start +
+	 * len - 1) / 2
+	 *
+	 * @param c     output array to accumulate column sums into
+	 * @param nRows number of rows, which is used because it is covered by the breakpoints
+	 */
+	/**
+	 * Accumulates the sum of all decompressed values across all columns into c[0].
+	 */
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Double comment openings

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 13]
```diff
@@ -0,0 +1,1443 @@
...
+		if(!isAddSub && !(op.fn instanceof Multiply || op.fn instanceof Divide))
+			throw new NotImplementedException("Unsupported binary op: " + op.fn.getClass().getSimpleName());
+
+		for(int col = 0; col < numCols; col++) {
+			double rowValue = v[_colIndexes.get(col)];
+			int numSegs = interceptsPerCol[col].length;
+			newIntercepts[col] = new double[numSegs];
+
+			// Plus/Minus: slope is translation-invariant, only intercept shifts
+			newSlopes[col] = isAddSub ? slopesPerCol[col].clone() : new double[numSegs];
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: A left intercept minus must invert the slope as well.
Division is not linear at all, so left division should not be supported in linear compressed.

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 14]
```diff
@@ -0,0 +1,1443 @@
...
+
+	private boolean colContainsValue(int col, double pattern) {
+		int[] breakpoints = breakpointsPerCol[col];
+		double[] intercepts = interceptsPerCol[col];
+		double[] slopes = slopesPerCol[col];
+		for(int seg = 0; seg < breakpoints.length - 1; seg++) {
+			int start = breakpoints[seg];
+			int len = breakpoints[seg + 1] - start;
+			if(len <= 0)
+				continue;
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Should not happen; otherwise it should be guarded when creating the data structure

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 15]
```diff
@@ -0,0 +1,1443 @@
...
+				// constant segment: all values equal b
+				if(Double.compare(b, pattern) == 0)
+					return true;
+				continue;
+			}
+
+			// check if pattern lies on the line: solve m*x + b = pattern for x
+			double x = (pattern - b) / m;
+			int xi = (int) x;
+			if(xi >= start && xi < start + len && Double.compare(m * xi + b, pattern) == 0)
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Exact comparison might be too strict here (as values). Maybe allow an error threshold similar to ColGroupLienarFunctional

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 16]
```diff
@@ -0,0 +1,1443 @@
...
+				return true;
+		}
+		return false;
+	}
+
+	private AColGroup decompress() {
+		IColIndex columns = ColIndexFactory.create(numRows);
+		MatrixBlock mb = new MatrixBlock(numRows, getNumCols(), false);
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Should correspond to num cols, not num rows

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 17]
```diff
@@ -0,0 +1,1443 @@
...
+		double[][] twoDimArray = new double[numRows][];
+		for(int i = 0; i < numRows; i++) {
+			int twoDimArray_lenght = in.readInt();
+			twoDimArray[i] = new double[twoDimArray_lenght];
+			for(int j = 0; j < twoDimArray_lenght; j++) {
+				twoDimArray[i][j] = in.readDouble();
+			}
+		}
+		return twoDimArray;
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Are these functions necessary? Couldn't we use existing utility (e.g, ColGroupIO.readDoubleArray or similar)? If not, maybe add utility function to ColGroupIO

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 18]
```diff
@@ -0,0 +1,1443 @@
...
+		double total = 0.0;
+		int[] bp = breakpointsPerCol[col];
+		double[] slopes = slopesPerCol[col];
+		double[] intercepts = interceptsPerCol[col];
+		for(int seg = 0; seg + 1 < bp.length; seg++) {
+			int start = bp[seg];
+			int end = bp[seg + 1];
+			int len = end - start;
+			if(len <= 0)
+				continue;
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Remove

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 19]
```diff
@@ -0,0 +1,1443 @@
...
+		return new ColGroupPiecewiseLinearCompressed(newColIndex, newBp, newSlopes, newIntercepts, numRows);
+	}
+
+	/**
+	 * piecewise linear groups encode all columns mathematically without sparse structural layouts. Returns null.
+	 */
+	@Override
+	public AColGroup removeEmptyColsSubset(IColIndex indexes, IntArrayList emptyCols) {
+		return null;
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:1256]: Should rather throw NotImplemented

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 20]
```diff
@@ -0,0 +1,1443 @@
...
+
+	/**
+	 * row subset filtering for piecewise linear functions requires full segment rebuilding. Returns null.
+	 */
+	@Override
+	public AColGroup removeEmptyRows(boolean[] emptyRows, int newNumRows) {
+		return null;
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:1264]: Throw

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 21]
```diff
@@ -0,0 +1,1443 @@
...
+
+	/**
+	 * piecewise linear representation is already order-dependent on row indices. Returns null.
+	 */
+	@Override
+	public AColGroup sort() {
+		return this;
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:1272]: Throw, only correct for already sorted data

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 22]
```diff
@@ -0,0 +1,1443 @@
...
+
+	/**
+	 * Piecewise linear column groups cannot be reduced to fewer columns; returns null.
+	 */
+	@Override
+	public AColGroup reduceCols() {
+		return null;
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:1280]: throw

File: src/test/java/org/apache/sysds/performance/PiecewiseLinearCompressionPerformanceTest.java
Comment (janniklinde) [inline src/test/java/org/apache/sysds/performance/PiecewiseLinearCompressionPerformanceTest.java:1]: Remove legacy text

File: src/test/java/org/apache/sysds/test/component/compress/colgroup/ColGroupPiecewiseLinearCompressedTest.java
Comment (janniklinde) [inline src/test/java/org/apache/sysds/test/component/compress/colgroup/ColGroupPiecewiseLinearCompressedTest.java:1]: Remove DP tests

File: src/test/java/org/apache/sysds/test/component/compress/colgroup/ColGroupPiecewiseLinearCompressedOperationsTest.java
Comment (janniklinde) [inline src/test/java/org/apache/sysds/test/component/compress/colgroup/ColGroupPiecewiseLinearCompressedOperationsTest.java:1]: A few recurring patterns I noticed here:

* Several tests only verify return types, non-null values, or that unsupported methods throw
* Some expected results derived from the implementation itself instead of being computed independently
* Error tolerances sometimes very loose relative to input range
* Missing edge cases (e.g., non-commutative ops, transposed input, zero target loss, serialization through the normal `ColGroupIO`)
* Leftover DP tests

**Review by janniklinde — CHANGES_REQUESTED** [review]
Thanks for the updates. I left comments in the code which should still be fixed. Most of them are minor and easy to patch.

Please address those remaining issues to pass the project.

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java (hunk 1 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java]: Done. Removed

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java (hunk 2 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java:100]: Done. Clear documentation comment above the function

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java (hunk 3 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/functional/PiecewiseLinearUtils.java]: Done

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java (hunk 19 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:1256]: Done

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java (hunk 20 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:1264]: Done

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java (hunk 21 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:1272]: Done

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java (hunk 22 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:1280]: Done

File: src/test/java/org/apache/sysds/performance/PiecewiseLinearCompressionPerformanceTest.java
Comment (Mancer1) [reply src/test/java/org/apache/sysds/performance/PiecewiseLinearCompressionPerformanceTest.java:1]: Done

File: src/test/java/org/apache/sysds/test/component/compress/colgroup/ColGroupPiecewiseLinearCompressedTest.java
Comment (Mancer1) [reply src/test/java/org/apache/sysds/test/component/compress/colgroup/ColGroupPiecewiseLinearCompressedTest.java:1]: Done

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupFactory.java (hunk 4 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupFactory.java]: Done

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java (hunk 7 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Done

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupFactory.java (hunk 5 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupFactory.java]: Done. Just need Tests for coverage

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java (hunk 18 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Done

**Comment by janniklinde** [comment]
Let me know once all issues have been addressed for me to make a final pass over the code @Mancer1.

**Comment by Mancer1** [comment]
@janniklinde, I'll tell you when we're fully done with the changes. I'm just replying "Done" to your comments for better team coordination and to know which changes have been made.

**Comment by Mancer1** [comment]
@janniklinde, we have made the changes you have requested. You may rerun the workflows.

**Comment by Mancer1** [comment]
@janniklinde, as mentioned before, the Java Code Style failure happens on files that are not part of our PR. We have done both dev/format-changed.sh and dev/format-changed.sh upstream/main before we pushed our code.

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 23]
```diff
@@ -0,0 +1,1390 @@
...
+/**
+ * This class represents a new ColGroup which is compresses column into segments (piecewise linear) to represent the
+ * original Data each column is approximate by a set of linear segments defined by breakpoints, slopes and intercepts
+ */
+
+public class ColGroupPiecewiseLinearCompressed extends AColGroupCompressed {
+	/**
+	 * breakpoints indices per column to define the segment boundaries slopes of the regression line per segment per
+	 * column intercepts of the regression line per segment per column
+	 */
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Weird comment, maybe one above class

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 24]
```diff
@@ -0,0 +1,1390 @@
...
+	int[][] breakpointsPerCol;
+	double[][] slopesPerCol;
+	double[][] interceptsPerCol;
+	int numRows;
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Why not private?

File: src/test/java/org/apache/sysds/performance/PiecewiseLinearCompressionPerformanceTest.java [hunk 25]
```diff
@@ -0,0 +1,203 @@
...
+import org.apache.sysds.runtime.matrix.operators.RightScalarOperator;
+import org.apache.sysds.runtime.matrix.operators.ScalarOperator;
+import org.apache.sysds.utils.stats.Timing;
+import org.apache.sysds.test.TestUtils;
+
+/**
+ * Performance benchmark for piecewise linear compression. Successive is benchmarked across large matrices to show
+ * scalability. DP is only used as a quality reference on small matrices due to quadratic complexity
+ *
+ */
```
Comment (janniklinde) [inline src/test/java/org/apache/sysds/performance/PiecewiseLinearCompressionPerformanceTest.java:39]: Outdated comment

File: src/test/java/org/apache/sysds/performance/PiecewiseLinearCompressionPerformanceTest.java [hunk 26]
```diff
@@ -0,0 +1,203 @@
...
+			double slope = 0.5 * (c + 1);
+			double intercept = 10.0 - c;
+			for(int r = 0; r < nr; r++) {
+				mb.set(r, c, slope * r + intercept);
+			}
+		}
+		return mb;
+	}
+
+	/// returns a average number of segments per column
```
Comment (janniklinde) [inline src/test/java/org/apache/sysds/performance/PiecewiseLinearCompressionPerformanceTest.java]: Unnecessary comment

File: src/test/java/org/apache/sysds/test/component/compress/colgroup/ColGroupPiecewiseLinearCompressedOperationsTest.java [hunk 27]
```diff
@@ -0,0 +1,1150 @@
...
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for ColGroupPiecewiseLinearCompressed operations.
+ */
```
Comment (janniklinde) [inline src/test/java/org/apache/sysds/test/component/compress/colgroup/ColGroupPiecewiseLinearCompressedOperationsTest.java]: Unnecessary

File: src/test/java/org/apache/sysds/test/component/compress/colgroup/ColGroupPiecewiseLinearCompressedOperationsTest.java [hunk 28]
```diff
@@ -0,0 +1,1150 @@
...
+			intercepts[s] = rng.nextDouble() * 4 - 2;
+		}
+		/// PLC Piecewise Linear Compressed
+		AColGroup colGroupPLC = ColGroupPiecewiseLinearCompressed.create(
+			ColIndexFactory.create(new int[] {rng.nextInt(20)}), new int[][] {breakpoints}, new double[][] {slopes},
+			new double[][] {intercepts}, rows);
+
+		assertTrue("disk size should be positive", colGroupPLC.getExactSizeOnDisk() > 0);
+		assertTrue("num values should be positive", colGroupPLC.getNumValues() > 0);
+	}
```
Comment (janniklinde) [inline src/test/java/org/apache/sysds/test/component/compress/colgroup/ColGroupPiecewiseLinearCompressedOperationsTest.java]: Weak assert

**Review by janniklinde — COMMENTED** [review]
I did another review pass and looks much better now. Most of the comments have been resolved. Have a look at the comments that are still open and the ones that I added now.

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupFactory.java (hunk 5 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupFactory.java]: A test with strong assertions has been implemented

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java (hunk 8 above)
Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java]: Why is this not resolved? I did remove the defensive copy.

**Comment by Mancer1** [comment]
@janniklinde, could you perhaps rerun just the functions.mlcontext check? In our main branch, this check came out as successful.

File: src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java [hunk 29]
```diff
@@ -0,0 +1,1387 @@
...
+		return agg;
+	}
+
+	/**
+	 * Two piecewise linear groups have the same index structure if they are both piecewise linear.
+	 */
+	@Override
+	public boolean sameIndexStructure(AColGroupCompressed that) {
+		return that instanceof ColGroupPiecewiseLinearCompressed;
+	}
```
Comment (janniklinde) [inline src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:792]: I think this is wrong, sameIndexStructure, simply return false

Comment (Mancer1) [reply src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupPiecewiseLinearCompressed.java:792]: @janniklinde, Done. Ready for the next workflow

**Comment by janniklinde** [comment]
LGTM. Thanks for your contribution
